### ISSUE TITLE
Could not find a declaration file for module 'styled-components'

### ISSUE THREAD

**Issue by plxel** [styled-components/styled-components#4235]
## Environment

`npx envinfo --system --binaries --npmPackages styled-components,babel-plugin-styled-components --markdown`

## System:
 - OS: macOS 13.5.2
 - CPU: (8) arm64 Apple M1 Pro
 - Memory: 119.91 MB / 32.00 GB
 - Shell: 5.9 - /bin/zsh
## Binaries:
 - Node: 18.18.0 - /usr/local/bin/node
 - npm: 9.8.1 - /usr/local/bin/npm
## npmPackages:
 - styled-components: ^6.1.2 => 6.1.2 

## Reproduction

https://codesandbox.io/p/sandbox/styled-components-with-ts-jqdp67?file=%2Fsrc%2FButton.tsx%3A4%2C17

## Steps to reproduce
![image](https://github.com/styled-components/styled-components/assets/1496959/55c357b6-37a9-4f16-99ec-d8be767c3968)

## Expected Behavior
No error occurs, types resolved

## Actual Behavior
```
Could not find a declaration file for module 'styled-components'. '<project-path>/node_modules/styled-components/dist/styled-components.cjs.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/styled-components` if it exists or add a new declaration (.d.ts) file containing `declare module 'styled-components';`
```

**Comment by Tbhesswebber** [comment]
For anyone who lands here after #4237 gets merged, you can verify that you're running into the same issues by looking in `node_modules/styled-components`.  If the `package.json` in that directory shows `"version": "6.1.2"` or higher and the `dist` directory doesn't have the declaration file, this is likely the issue you're hitting.

![image](https://github.com/styled-components/styled-components/assets/28069638/3a3e418d-79db-495c-9436-819af4c445a0)

**Comment by AHorak** [comment]
I have the same problem!

**Comment by chanho-Yoon** [comment]
I have the same problem

**Comment by quantizor** [comment]
Sorry about this folks, when I switched the pm to bun there was a small amount of dependency drift which pulled in a rollup plugin update that quietly broke declaration file generation. The issue has been solved with 6.1.3 (thank you @benjaminrae and contributors) and I've gone ahead and pinned all dependencies going forward to exact versions.
